File: app/src/main/java/com/tunjid/rcswitchcontrol/dialogfragments/RenameSwitchDialogFragment.java
package com.tunjid.rcswitchcontrol.dialogfragments;

import android.annotation.SuppressLint;
import android.app.Dialog;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.EditText;

import com.tunjid.rcswitchcontrol.R;
import com.tunjid.rcswitchcontrol.model.RcSwitch;

import java.util.Objects;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.fragment.app.DialogFragment;
import androidx.fragment.app.FragmentActivity;


@SuppressLint("InflateParams")
public class RenameSwitchDialogFragment extends DialogFragment {

    private static final String SWITCH = "SWITCH";

    private RcSwitch rcSwitch;

    public static RenameSwitchDialogFragment newInstance(RcSwitch rcSwitch) {

        RenameSwitchDialogFragment fragment = new RenameSwitchDialogFragment();
        Bundle args = new Bundle();
        args.putParcelable(SWITCH, rcSwitch);
        fragment.setArguments(args);
        return fragment;
    }

    public RenameSwitchDialogFragment() {

    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        rcSwitch = Objects.requireNonNull(getArguments()).getParcelable(SWITCH);
    }

    @Override
    @NonNull
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        final SwitchNameListener listener = ((SwitchNameListener) getParentFragment());

        FragmentActivity activity = requireActivity();
        final LayoutInflater inflater = activity.getLayoutInflater();

        final View view = inflater.inflate(R.layout.dialog_rename_switch, null);
        final EditText editText = view.findViewById(R.id.switch_name);

        editText.setText(rcSwitch.getName());

        return new AlertDialog.Builder(activity).setTitle(R.string.rename_switch)
                .setView(view)
                .setPositiveButton(R.string.rename, (dialog, id) -> {
                    rcSwitch.setName(editText.getText().toString());
                    Objects.requireNonNull(listener).onSwitchRenamed(rcSwitch);
                    dismiss();
                })
                .setNegativeButton(R.string.cancel, (dialog, id) -> dismiss())
                .create();
    }

    public interface SwitchNameListener {
        void onSwitchRenamed(RcSwitch rcSwitch);
    }
}

File: app/src/main/java/com/tunjid/rcswitchcontrol/dialogfragments/RenameSwitchDialogFragment.kt
/*
 * MIT License
 *
 * Copyright (c) 2019 Adetunji Dahunsi
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package com.tunjid.rcswitchcontrol.dialogfragments

import android.annotation.SuppressLint
import android.app.Dialog
import android.os.Bundle
import androidx.fragment.app.DialogFragment
import androidx.fragment.app.activityViewModels
import com.rcswitchcontrol.protocols.Name
import com.rcswitchcontrol.protocols.renamePayload
import com.tunjid.androidx.core.delegates.fragmentArgs
import com.tunjid.rcswitchcontrol.R
import com.tunjid.rcswitchcontrol.viewmodels.ControlViewModel

@SuppressLint("InflateParams")
class RenameSwitchDialogFragment : DialogFragment() {

    private var name by fragmentArgs<Name>()
    private val viewModel by activityViewModels<ControlViewModel>()

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog = editTextDialog { editText, builder ->
        editText.setText(name.value)

        builder
            .setTitle(R.string.rename_switch)
            .setPositiveButton(R.string.rename) { _, _ ->
                viewModel.dispatchPayload(name.copy(value = editText.text.toString()).renamePayload)
                dismiss()
            }
    }

    companion object {
        fun newInstance(name: Name): RenameSwitchDialogFragment = RenameSwitchDialogFragment().apply {
            this.name = name
        }
    }
}
